### PR TITLE
[SDK] Fix withItem loop

### LINKED ISSUE #2570: ParallelFor parameter parsing does not retain user input format
When running a pipeline with a ParallelFor condition over a list of dictionaries, camelCase and snake_case inputs are altered.

This won't work:

    with kfp.dsl.ParallelFor([{"helloWorld": "blah"}]) as item:
        my_op(item.helloWorld)

    with kfp.dsl.ParallelFor([{"hello_world": "blah"}]) as item:
        my_op(item.hello_world)

If I am reading this correctly, then the `my_op` is parsing `item.hello_world` as `item.hello-world` (and `helloWorld` as `helloworld`. However, the ParallelFor operation isn't  performing the same parsing. (I presume it is delivering `item.hello_world`, exactly as written in the dictionary, but I cannot confirm this).

The following works correctly:

    with kfp.dsl.ParallelFor([{"helloworld": "blah"}]) as item:
        my_op(item.helloWorld)

    with kfp.dsl.ParallelFor([{"hello-world": "blah"}]) as item:
        my_op(item.hello_world)

### PR DESCRIPTION
**Pull request by numerology**
Partially fix #2570 

Question: I can imagine for withparams we have the same problem if user provides arbitrary names in the loop variable's runtime value. How to prevent that from happening? Shall we bypass k8s name sanitization in this case?

---
This change is [Reviewable](https://reviewable.io/reviews/kubeflow/pipelines/2572)

### CODE REVIEW

**Comment by Ark-kun** [comment]
> we have the same problem if user provides arbitrary names in the loop variable's runtime value

There might be a problem if the user has non-Kubernetes keys in their dictionaries since the `{{item.*}}` placeholders might be converted. The runtime values won't be a problem. But the placeholder might not be able to access them (e.g. you cannot access keys with spaces).

File: sdk/python/kfp/compiler/compiler.py [hunk 1]
```diff
@@ -487,7 +487,16 @@ def _group_to_dag_template(self, group, inputs, outputs, dependencies):
 
           task['withParam'] = withparam_value
         else:
-          task['withItems'] = sub_group.loop_args.to_list_for_task_yaml()
+          # Need to sanitize the dict keys for consistency.
+          loop_tasks = sub_group.loop_args.to_list_for_task_yaml()
+          isinstance(loop_tasks[0], dict)
```
Comment (Ark-kun) [inline sdk/python/kfp/compiler/compiler.py]: I'm not sure this line does anything.

**Comment by numerology** [comment]
> > we have the same problem if user provides arbitrary names in the loop variable's runtime value
> 
> There might be a problem if the user has non-Kubernetes keys in their dictionaries since the `{{item.*}}` placeholders might be converted. The runtime values won't be a problem. But the placeholder might not be able to access them (e.g. you cannot access keys with spaces).

That's true. So shall we alert the user whenever they want to access a non-Kubernetes keys in their ContainerOp, for example, try to use `item.a_a`, which will be converted to `a-a` under the hood?

File: sdk/python/kfp/compiler/compiler.py [hunk 2]
```diff
@@ -487,7 +487,16 @@ def _group_to_dag_template(self, group, inputs, outputs, dependencies):
...
+          sanitized_tasks = []
+          for item in loop_tasks:
```
Comment (Ark-kun) [inline sdk/python/kfp/compiler/compiler.py]: Nit: `for argument_set in loop_tasks:`

**Comment by Ark-kun** [comment]
> So shall we alert the user whenever they want to access a non-Kubernetes keys in their ContainerOp, for example, try to use `item.a_a`, which will be converted to `a-a` under the hood?

That would be great. But there is a problem - python requires underscores while kubernetes requires dashes. So the names will be converted often.

Update. I just remember that this is not Kubernetes limitation, but rather Argo limitation. And Argo lifted it in 2.3.0. So maybe we should just stop converting underscores to dashes. Then the problem will mostly go away.

**Comment by numerology** [comment]
> > So shall we alert the user whenever they want to access a non-Kubernetes keys in their ContainerOp, for example, try to use `item.a_a`, which will be converted to `a-a` under the hood?
> 
> That would be great. But there is a problem - python requires underscores while kubernetes requires dashes. So the names will be converted often.
> 
> Update. I just remember that this is not Kubernetes limitation, but rather Argo limitation. And Argo lifted it in 2.3.0. So maybe we should just stop converting underscores to dashes. Then the problem will mostly go away.

Thanks for the info! If that's the case, I'd prefer the latter way, because as discussed in the related issue, current fix cannot fix the problem with withParams. I think it would be better off if we just do not use sanitized name in this case.

Let me try to improve it a little bit and will let you know when I'm done.

**Comment by Ark-kun** [comment]
> Let me try to improve it a little bit and will let you know when I'm done.

It's fine to do that in another PR though. It might require some sizable amount of work and testing.

**Comment by numerology** [comment]
> > Let me try to improve it a little bit and will let you know when I'm done.
> 
> It's fine to do that in another PR though. It might require some sizable amount of work and testing.

Okay.

**Pull request merged**